Bugfix: BB-159 version delete not producing notification

welansari opened the pull request:
Issue: [BB-159](https://scality.atlassian.net/browse/BB-159)

**Original Issue:**
Version delete events not producing notifications as we only process master object events.
Simply processing both master and version events does fix the issue but introduces another, where now we have two notifications for the same event in certain cases, like when we delete the latest version of an object.

After the work done [in this PR](https://github.com/scality/Arsenal/pull/1944), we now update the objects' metadata before deleting them, adding a new event to the oplog containing the latest metadata of the object. The update adds a deletion flag to the object making it simple to distinguish objects that are in the process of deletion.

We now totally ignore actual deletion events and only process the update event that was added, as we do need the object metadata for the notifications.

The following are the cases that are processed:
- In non versioned buckets, we process the master object events.
- In versioned buckets, we only process version events and ignore all master events, this makes us avoid dealing with PHD cases.
- In versioning suspended buckets, we process both version and master events. As the master object is considered an independent version.

**Other Changes:**
- Removed version id from the object key in the notifications
- Set version id field to null in the notification when the object is a null version

jonathan-gramain commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -38,6 +39,45 @@ class NotificationQueuePopulator extends QueuePopulatorExtension {
+            versioningConfig.Status === 'Suspended';
+        // bucket versioning active condition (accept versions)
+        const versionActiveCond = versioningConfig &&
+            versioningConfig.Status === 'Enabled' && !isMasterKey(key);
```
Beware, as versioned buckets can contain objects with only a master key and no version key, for objects written before versioning was enabled. For notification purpose it may not matter much but we may lose notifications around the time a bucket's versioning is enabled due to this. Can you check this to make sure this is wanted?

Maybe we would need to process all keys and ignore versions when we see a master instead.

francoisferrand replied:
is there an easy way to identify the "last" version of an object?
just ignoring changes to this last version (and processing master instead) would deal with all situations I guess...

welansari replied:
Not that i know of, i checked the metadata we store for versions and none contain that information

jonathan-gramain replied:
the "last" version is the first version key listed for a particular object key. The latest version ID is also stored in the master key if it helps.

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/backbeat/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




There is a conflict between your branch `bugfix/BB-159-version-delete-not-producing-notification` and the
destination branch `development/8.4`.

Please resolve the conflict on **the feature branch** (`bugfix/BB-159-version-delete-not-producing-notification`).

```sh
git fetch && \
git checkout origin/bugfix/BB-159-version-delete-not-producing-notification && \
git merge origin/development/8.4
```

Resolve merge conflicts and commit

```sh
git push origin HEAD:bugfix/BB-159-version-delete-not-producing-notification
```

bert-e commented:
# Branches have diverged



This pull request's source branch `bugfix/BB-159-version-delete-not-producing-notification` has diverged from
`development/8.4` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/8.4` into `bugfix/BB-159-version-delete-not-producing-notification`
* Rebase `bugfix/BB-159-version-delete-not-producing-notification` onto `origin/development/8.4`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-159 contains:


* `7.10.4`

* `8.4.21`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.21`


Please check the `Fix Version/s` of BB-159, or the target
branch of this pull request.

welansari commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-159 contains:


* `7.10.4`

* `8.4.21`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.21`


Please check the `Fix Version/s` of BB-159, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-159 contains:


* `8.4.23`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.24`

* `8.5.1`


Please check the `Fix Version/s` of BB-159, or the target
branch of this pull request.

francoisferrand commented on lib/queuePopulator/LogReader.js:
```diff
@@ -379,20 +379,8 @@ class LogReader {
+            if (entry.value === undefined) {
                 return null;
             }
             return ext.filter({
```
if we detect this 'update' is actually a 'delete', should we not update the record, so that downstream extensions receive a 'delete' as before?

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-159 contains:


* `8.4.23`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.25`

* `8.5.2`


Please check the `Fix Version/s` of BB-159, or the target
branch of this pull request.

Commit:
bugfix: BB-159 use update operation performed before delete

Before each objet deletion we first perform an update of the object's
metadata putting a deletion flag into them. We now use this update
operation as a delete event, and ignore the actual delete as it contains
no metadata.

Commit:
bugfix: BB-159 update notificationQueuePopulator to use the new update event

The update operation we do before deleting objects will be used as a deletion
event. This helps determine wether an object is versioned, non versioned or
version suspended.

To avoid pushing two notifications for the same events we ignore all versioned
masters and only process the versions. The only cases where master events are
processed is when the master is non versioned or when it's a null version

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/BB-159-version-delete-not-producing-notification` into
`development/8.4`
* [w/8.5/bugfix/BB-159-version-delete-not-producing-notification](https://github.com/scality/backbeat/compare/development/8.5...w/8.5/bugfix/BB-159-version-delete-not-producing-notification) will be merged into `development/8.5` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

Commit:
bugfix: BB-159 bump arsenal to 8.1.73

bert-e commented:
# History mismatch



Merge commit #3cbedfa30f87f0daff01cb2ae2a63d0716175bf3 on the integration branch
`w/8.5/bugfix/BB-159-version-delete-not-producing-notification` is merging a branch which is neither the current
branch `bugfix/BB-159-version-delete-not-producing-notification` nor the development branch
`development/8.5`.

It is likely due to a rebase of the branch `bugfix/BB-159-version-delete-not-producing-notification` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

welansari commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/BB-159-version-delete-not-producing-notification` into
`development/8.4`
* [w/8.5/bugfix/BB-159-version-delete-not-producing-notification](https://github.com/scality/backbeat/compare/development/8.5...w/8.5/bugfix/BB-159-version-delete-not-producing-notification) will be merged into `development/8.5` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

welansari commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/8.3`



Please check the status of the associated issue BB-159.

Goodbye kerkesni.
